<div style="text-align: center;">
  <img src="https://github.com/Hack-io-Data/Imagenes/blob/main/01-LogosHackio/logo_naranja@4x.png?raw=true" alt="esquema" />
</div>

# Laboratorio Filtrado de Datos

En este laboratorio usaremos el DataFrame creado en la lección anterior resultado de la unión de los conjuntos de datos compartidos contigo. 

**Instrucciones:**

1. Lee cuidadosamente el enunciado de cada ejercicio.

2. Implementa la solución en la celda de código proporcionada.

In [25]:
# Tratamiento de datos
# -----------------------------------------------------------------------
import pandas as pd
import numpy as np


# Configuración
# -----------------------------------------------------------------------
pd.set_option('display.max_columns', None) # para poder visualizar todas las columnas de los DataFrames

# antes de empezar vamos a leer los datos
df = pd.read_csv("datos/netflix_titles_originales_merge.csv", index_col=0)
df.head(1)



Title        Genre         Premiere  Runtime  IMDB Score  \
0  Dick Johnson Is Dead  Documentary  October 2, 2020     90.0         7.5   

  Language show_id   type                 title         director cast  \
0  English      s1  Movie  Dick Johnson Is Dead  Kirsten Johnson  NaN   

         country          date_added  release_year rating duration  \
0  United States  September 25, 2021          2020  PG-13   90 min   

       listed_in                                        description  
0  Documentaries  As her father nears the end of his life, filmm...

#### Ejercicio 1: ¿Cuáles son todos los títulos lanzados en 2021 que son del tipo 'Movie'?

In [ ]:
df_filter_1 = df["release_year"] == 2021
df_filter_1A = df["type"] == "Movie"

df_1 = df[df_filter_1 & df_filter_1A]

df_1.count()


#### Ejercicio 2: ¿Qué títulos tienen una duración de 90 minutos? Muestra solo las columnas de título y de duración.

In [29]:
df_filter_2 = df["duration"] == "90 min"

df_2 = df[["title","duration"]][df_filter_2].reset_index()

df_2

index                            title duration
0        0             Dick Johnson Is Dead   90 min
1       94                        Show Dogs   90 min
2      107                 A Champion Heart   90 min
3      229                     The Old Ways   90 min
4      283             Misha and the Wolves   90 min
..     ...                              ...      ...
147   8651                            Tyson   90 min
148   8655             Unaccompanied Minors   90 min
149   8670                    Vantage Point   90 min
150   8735  Who's That Knocking at My Door?   90 min
151   8776                       Yellowbird   90 min

[152 rows x 3 columns]

#### Ejercicio 3: ¿Cuáles son todos los títulos del país 'India' lanzados después de 2018?

In [ ]:
df_filter_3 = df["release_year"] > 2018
df_filter_3a = df["country"] == "India"

df_3 = df[df_filter_3 & df_filter_3a]

df_3.count()

#### Ejercicio 4: ¿Qué títulos tienen clasificación 'TV-14' o 'R'?

In [ ]:
# primero definimos una lista con los valores que queremos filtrar
lista_valores_4 = ['TV-14', 'R']

# aplicamos el método 'isin()' pasando la lista
df_clasificacion_4 = df[df["rating"].isin(lista_valores_4)]
df_clasificacion_4.count()

#### Ejercicio 5: ¿Cuáles son los títulos dirigidos por 'Christopher Nolan' o 'Martin Scorsese'? ¿Cuántos a dirigido cada uno de los directores?

In [59]:
# primero definimos una lista con los valores que queremos filtrar
lista_valores_5 = ['Christopher Nolan', 'Martin Scorsese']

# aplicamos el método 'isin()' pasando la lista
df_clasificacion_5 = df[df["director"].isin(lista_valores_5)]

df_clasificacion_5A = df_clasificacion_5[df_clasificacion_5["director"] == 'Christopher Nolan'].reset_index()

df_clasificacion_5B = df_clasificacion_5[df_clasificacion_5["director"] == 'Martin Scorsese'].reset_index()

print(f"Peliculas de Christopher Nolan: {df_clasificacion_5A.shape[0]}")

print(f"Peliculas de Martin Scorsese: {df_clasificacion_5B.shape[0]}")

Peliculas de Christopher Nolan: 1
Peliculas de Martin Scorsese: 12


#### Ejercicio 6: ¿Qué títulos pertenecen a los países 'United States', 'United Kingdom' o 'Canada'? Dentro de estos países, ¿cuántas películas tienen más de un lenguaje?

#### Ejercicio 7: ¿Qué títulos fueron lanzados entre 2015 y 2020, inclusive?

#### Ejercicio 8: ¿Cuáles son los títulos cuya puntuación en IMDb está entre 6.5 y 8.0 y que no sepamos cual es su duración?

#### Ejercicio 9: ¿Cuáles son las películas cuya duración de 90 y 150 minutos?

#### Ejercicio 10: ¿Cuáles son los títulos que contienen la palabra 'Love' en su título y que sean de tipo TV Show?

#### Ejercicio 11: ¿Cuáles son los títulos en los que el director tiene el apellido 'Johnson'?

#### Ejercicio 12: ¿Tenemos algún documental ('Documentaries') de Mike Tyson?

#### Ejercicio 13: Selecciona solo las columnas 'title', 'country' y 'release_year' del dataset.

#### Ejercicio 14: Filtra las columnas que contienen la palabra 'date'.

#### Ejercicio 15: Usando `query()`, filtra todos los títulos del país 'France' con una duración de 80 minutos.

#### Ejercicio 16: Filtra los títulos que son películas ('Movie') y fueron lanzados antes de 2010.

#### Ejercicio 17: Encuentra los 10 títulos con las puntuaciones más altas en IMDb.

#### Ejercicio 18: Encuentra los 5 títulos con las puntuaciones más bajas en IMDb.

#### Ejercicio 19: Filtra todos los títulos que no tienen un director asignado (valores nulos en la columna 'director').

#### Ejercicio 20: Filtra todos los títulos que tienen información completa en las columnas 'title', 'country' y 'IMDB Score'.

#### Ejercicio 21: Identifica si hay títulos duplicados en la columna 'title'.